### DISCUSSION TITLE
blog/preserve-source-ip-in-istio/

### DISCUSSION THREAD

**Discussion by giscus** [rootsongjc/rootsongjc.github.io#340] [Announcements]
# blog/preserve-source-ip-in-istio/

本文专注于如何在 Istio 服务网格中保持客户端源 IP 的透明性。

https://jimmysong.io/blog/preserve-source-ip-in-istio/

**Comment by kanlac** [comment]
谢谢博主的分享，文章质量很高，不过暂时有几个问题暂时还不太明白：
- `externalTrafficPolicy: Local` 这个设置似乎相当于是直接打到节点本地的 pod，如果环境中并没有外部负载均衡器，是不是就没有意义了？（因为内部的 gateway 需要做负载均衡）
- 文章中没有提到 UDP 流量的解决方案，如果想实现 UDP 基于客户端 IP 的亲和性，博主有推荐的技术方案吗？

**Reply by rootsongjc** [reply]
你提出了两个很好的问题，我来逐一回答：

1. 对于 `externalTrafficPolicy: Local` 设置，如果没有外部负载均衡器，是否还有意义？

在没有外部负载均衡器的情况下，实际上 `externalTrafficPolicy: Local` 设置并没有太大意义。因为在没有外部负载均衡器的情况下，流量直接进入集群，不会有 SNAT 操作，源 IP 本来就不会丢失。但是即使设置了这个选项，也不会有任何影响。

2. 对于 UDP 流量，如何实现基于客户端 IP 的亲和性？

Envoy 不支持使用 X-Forwarded-For 标头传递 UDP 流量的源 IP 信息。不过你可以使用 Proxy Protocol，它可以在 UDP 报文头中传递源 IP 信息。另一种方法是使用基于 IP 和端口的五元组哈希（源 IP 地址、源端口、目标 IP 地址、目标端口和协议），将 UDP 流量分配到同一目标 Pod 上，以实现亲和性。

**Reply by kanlac** [reply]
谢谢解答。

> 因为在没有外部负载均衡器的情况下，流量直接进入集群，不会有 SNAT 操作，源 IP 本来就不会丢失。

我在实际应用中发现用 Ingress-Nginx 转发 UDP 流量，发现数据包的源 IP 确实丢失了，可能 Ingress-Nginx 针对四层流量的转发策略不一样，会经过 SNAT 吧。

> 另一种方法是使用基于 IP 和端口的五元组哈希（源 IP 地址、源端口、目标 IP 地址、目标端口和协议），将 UDP 流量分配到同一目标 Pod 上，以实现亲和性。

我们最终采用了类似你说的这个方法，为 Ingress-Nginx 控制器编写扩展负载均衡脚本，使用客户端 IP 做一致性哈希，成功解决了 UDP 的 CIP 亲和性问题 ^_^
